File: 프로그래머스/lv2/이하은/하노이의 탑.kt
typealias PII = Pair<Int, Int>

class Solution {
    fun solution(n: Int): Array<IntArray> {
        var answer = mutableListOf<PII>()
        
        hanoi(n, 1, 3, 2, answer)
        
        return Array(answer.size) { i -> 
            intArrayOf(answer[i].first, answer[i].second)
        }
    }
    
    fun hanoi(n: Int, start: Int, end: Int, mid: Int, answer: MutableList<PII>) {
        if (n == 1) {
            answer.add(start to end)
            return 
        }
        
        // 가장 큰 원판 제외한 n-1개의 원판을 가운데로 옮긴다. 
        hanoi(n - 1, start, mid, end, answer)
        
        // 가장 큰 원판을 타겟으로 옮긴다. 
        answer.add(start to end)
        
        // 가운데 있던 원판들을 타겟으로 옮긴다. 
        hanoi(n - 1, mid, end, start, answer)
    }
}

File: 프로그래머스/lv2/오하늘/하노이의 탑.py
def dfs(n, start, end, sub, answer):
    if n == 1:
        answer.append([start, end]) # 1개
    else:
        dfs(n-1, start, sub, end, answer) # 2번 기둥으로
        answer.append([start, end]) # 가장 큰 원반을 끝으로
        dfs(n-1, sub, end, start, answer) # 2->3번으로


def solution(n):
    answer = []
    dfs(n, 1, 3, 2, answer)
    
    return answer